### PR TITLE
Add asyncTriggerTimeout method

### PR DESCRIPTION
**Pull request by jacobp100**
## Summary

Adds method to terminate running JS without using timers (see #929)

## Test Plan

Unsure

### CODE REVIEW

File: API/hermes/hermes.h [hunk 1]
```diff
@@ -188,12 +188,16 @@ class HERMES_EXPORT HermesRuntime : public jsi::Runtime {
...
   void unregisterForProfiling();
 
-  /// Register this runtime for execution time limit monitoring, with a time
-  /// limit of \p timeoutInMs milliseconds.
+  /// Asynchronously terminates the current execution.
   /// All JS compiled to bytecode via prepareJS, or evaluateJS, will support the
   /// time limit monitoring.  If JS prepared in other ways is executed, care
   /// must be taken to ensure that it is compiled in a mode that supports the
   /// monitoring (i.e., the emitted code contains async break checks).
+  void notifyNeedsTermination();
```
Comment (neildhar) [inline API/hermes/hermes.h:196]: I'd suggest renaming this to `asyncTriggerTimeout`, to make it clear that this will throw a timeout exception.

File: API/hermes/hermes.h [hunk 2]
```diff
@@ -188,12 +188,16 @@ class HERMES_EXPORT HermesRuntime : public jsi::Runtime {
   /// Unregister this runtime for sampling profiler.
   void unregisterForProfiling();
 
-  /// Register this runtime for execution time limit monitoring, with a time
-  /// limit of \p timeoutInMs milliseconds.
+  /// Asynchronously terminates the current execution.
```
Comment (neildhar) [inline API/hermes/hermes.h:191]: We should also mention that this may be invoked from any thread (most other API methods may not).

**Review by neildhar — COMMENTED** [review]
Apologies for the delay in getting back to you. The PR looks good, I just have some comments regarding naming.

It would also be great if you could add a test for the new mechanism to verify it works. You can use https://github.com/facebook/hermes/blob/3a582a557f66ea1d82211187f793ed1111faefda/unittests/API/APITest.cpp#L376 as a reference, although it can be much shorter (it will need to spawn a thread though).

**Comment by jacobp100** [comment]
@neildhar Sorry I took so long to fix the issue you pointed out. Hopefully it's working now. Appreciate the comments, C++ is still very new to me

File: unittests/API/APITest.cpp [hunk 3]
```diff
@@ -433,6 +433,24 @@ TEST(HermesWatchTimeLimitTest, WatchTimeLimit) {
...
+  const char *forEver = "for (;;){}";
+  uint32_t ShortTimeoutMS = 123;
+  {
+    // Single runtime with ~20 minute limit that will not be reached.
+    auto rt = makeHermesRuntime();
+    std::thread t([&]() {
+      std::this_thread::sleep_for(std::chrono::milliseconds(ShortTimeoutMS));
+      rt->asyncTriggerTimeout();
+    });
+    t.detach();
```
Comment (neildhar) [inline unittests/API/APITest.cpp:447]: Why detach the thread here?

File: unittests/API/APITest.cpp [hunk 4]
```diff
@@ -433,6 +433,24 @@ TEST(HermesWatchTimeLimitTest, WatchTimeLimit) {
   }
 }
 
+TEST(HermesTriggerAsyncTimeoutTest, TriggerAsyncTimeout) {
+  // Some code that exercies the async break checks.
+  const char *forEver = "for (;;){}";
+  uint32_t ShortTimeoutMS = 123;
+  {
+    // Single runtime with ~20 minute limit that will not be reached.
```
Comment (neildhar) [inline unittests/API/APITest.cpp:441]: Looks like this comment needs to be updated

File: unittests/API/APITest.cpp [hunk 5]
```diff
@@ -433,6 +433,24 @@ TEST(HermesWatchTimeLimitTest, WatchTimeLimit) {
   }
 }
 
+TEST(HermesTriggerAsyncTimeoutTest, TriggerAsyncTimeout) {
+  // Some code that exercies the async break checks.
```
Comment (neildhar) [inline unittests/API/APITest.cpp:437]:
```suggestion
  // Some code that loops forever to exercise the async interrupt.
```

File: unittests/API/APITest.cpp (hunk 3 above)
Comment (jacobp100) [reply unittests/API/APITest.cpp:447]: I don’t know. What’s the alternative?

Comment (jacobp100) [reply unittests/API/APITest.cpp:447]: The tests failed without it. My understanding is `.join` won’t work because it’ll wait for the thread to finish - so we’ll end up in deadlock. But I didn’t actually try it so I could be wrong

File: unittests/API/APITest.cpp (hunk 4 above)
Comment (jacobp100) [inline unittests/API/APITest.cpp:441]:
```suggestion
```

File: unittests/API/APITest.cpp (hunk 3 above)
Comment (neildhar) [reply unittests/API/APITest.cpp:446]: We should be able to `join` after the `ASSERT_THROW`.

Comment (jacobp100) [reply unittests/API/APITest.cpp]: You're right - that works. What happened here? Without `join` and `detach`, it did not work. But to my understanding, the `ASSERT_THROW` won't return unless the thread executes the code to stop the runtime. But by that time, the `.join` shouldn't do anything, since we basically know the thread code has run. How does the `.join` afterwards make any impact here?

Comment (neildhar) [reply unittests/API/APITest.cpp]: From the reference for `joinable` ([link](https://en.cppreference.com/w/cpp/thread/thread/joinable)):

> A thread that has finished executing code, but has not yet been joined is still considered an active thread of execution and is therefore joinable.

And for `~thread` ([link](https://en.cppreference.com/w/cpp/thread/thread/~thread)):

> If `*this` has an associated thread (`joinable() == true`), `std::terminate()` is called.

So even if the thread has finished executing, it should be joined before it is destroyed. That said, that there is actually no guarantee that the thread has actually finished executing. All we know when the exception is thrown is that the interrupt flag in the runtime was set. In theory, the thread could have set the flag, and then immediately been suspended before it could return.

File: API/hermes/hermes.h [hunk 6]
```diff
@@ -188,12 +188,19 @@ class HERMES_EXPORT HermesRuntime : public jsi::Runtime {
   /// Unregister this runtime for sampling profiler.
   void unregisterForProfiling();
 
+  /// Define methods to interrupt JS execution and set time limits.
+  /// All JS compiled to bytecode via prepareJS, or evaluateJS, will support interruption and
+  /// time limit monitoring.  If JS prepared in other ways is executed, care
```
Comment (neildhar) [inline API/hermes/hermes.h:193]: Turns out that this isn't true, which is why the tests in release mode are failing in CircleCI. Async break check is actually disabled by default, and in release builds, with debugging disabled, they are not emitted at all.

We should fix this, but I don't want to block your PR on it, so let's make the requirements clear.
```suggestion
  /// All JS compiled to bytecode via prepareJS, or evaluateJS, will support interruption and
  /// time limit monitoring if the runtime is configured with AsyncBreakCheckInEval.  If JS prepared in other ways is executed, care
```

File: unittests/API/APITest.cpp [hunk 7]
```diff
@@ -433,6 +433,23 @@ TEST(HermesWatchTimeLimitTest, WatchTimeLimit) {
   }
 }
 
+TEST(HermesTriggerAsyncTimeoutTest, TriggerAsyncTimeout) {
+  // Some code that loops forever to exercise the async interrupt.
+  const char *forEver = "for (;;){}";
+  uint32_t ShortTimeoutMS = 123;
+  {
+    auto rt = makeHermesRuntime();
```
Comment (neildhar) [inline unittests/API/APITest.cpp:441]: For now, we need to manually enable this.

```suggestion
    auto rt = makeHermesRuntime(hermes::vm::RuntimeConfig::Builder().withAsyncBreakCheckInEval(true).build());
```

**Review by neildhar — CHANGES_REQUESTED** [review]
LGTM overall, this should fix the test failure and make it ready to merge
